In [4]:
import pandas as pd
import os

In [5]:
df = pd.read_csv('metaData.csv', na_values='-') #Let's begin by reading the data from the metaData.csv file. We replace the '-' with NaN values.
df.head()

name  busNumber  \
0  B183_2019-04-30_03-18-56_2019-04-30_08-44-20        183   
1  B183_2019-04-30_13-22-07_2019-04-30_17-54-02        183   
2  B183_2019-05-01_05-58-51_2019-05-01_22-32-30        183   
3  B183_2019-05-03_02-50-21_2019-05-03_05-53-20        183   
4  B183_2019-05-03_15-41-57_2019-05-03_23-06-24        183   

          startTime_iso  startTime_unix           endTime_iso  endTime_unix  \
0  2019-04-30T03:18:56Z      1556594336  2019-04-30T08:44:20Z    1556613860   
1  2019-04-30T13:22:07Z      1556630527  2019-04-30T17:54:02Z    1556646842   
2  2019-05-01T05:58:51Z      1556690331  2019-05-01T22:32:30Z    1556749950   
3  2019-05-03T02:50:21Z      1556851821  2019-05-03T05:53:20Z    1556862800   
4  2019-05-03T15:41:57Z      1556898117  2019-05-03T23:06:24Z    1556924784   

   drivenDistance busRoute  energyConsumption  itcs_numberOfPassengers_mean  \
0        77213.87      NaN       4.785852e+08                      5.538860   
1        59029.60       31       4.022585e+08                     33.114580   
2       240900.40       33       1.445733e+09                     19.689140   
3        42565.48      NaN       2.819867e+08                      1.685185   
4       125277.20       72       6.207258e+08                     23.753570   

   itcs_numberOfPassengers_min  itcs_numberOfPassengers_max  \
0                            0                           20   
1                            4                           74   
2                            0                           55   
3                            0                            8   
4                            1                           67   

   status_gridIsAvailable_mean  temperature_ambient_mean  \
0                     0.740640                  282.3780   
1                     0.855234                  287.5443   
2                     0.777860                  288.7490   
3                     0.767122                  282.4129   
4                     0.907342                  284.7325   

   temperature_ambient_min  temperature_ambient_max  
0                   281.15                   293.15  
1                   285.15                   293.15  
2                   280.15                   294.15  
3                   281.15                   292.15  
4                   282.15                   287.15

In [5]:
df.dtypes

name                             object
busNumber                         int64
startTime_iso                    object
startTime_unix                    int64
endTime_iso                      object
endTime_unix                      int64
drivenDistance                  float64
busRoute                         object
energyConsumption               float64
itcs_numberOfPassengers_mean    float64
itcs_numberOfPassengers_min       int64
itcs_numberOfPassengers_max       int64
status_gridIsAvailable_mean     float64
temperature_ambient_mean        float64
temperature_ambient_min         float64
temperature_ambient_max         float64
dtype: object

# 1. Extract all trips with `busRoute` 83.

### How many routes are in the dataset?

In [6]:
df.groupby('busRoute').size()

busRoute
31     17
32     26
33    155
46    123
72    158
83    846
N1     16
N2     39
N4     18
dtype: int64

### Let's only pick the trips with busRoute equal to 83.

In [7]:
df[df['busRoute']=='83']

name  busNumber  \
154   B183_2020-03-03_04-42-38_2020-03-03_19-44-51        183   
155   B183_2020-03-06_04-53-23_2020-03-06_19-44-42        183   
157   B183_2020-03-09_14-16-13_2020-03-09_19-34-17        183   
158   B183_2020-03-10_04-50-03_2020-03-10_19-51-25        183   
159   B183_2020-03-12_04-56-41_2020-03-12_19-44-57        183   
...                                            ...        ...   
1399  B208_2022-11-30_04-47-53_2022-11-30_19-50-22        208   
1400  B208_2022-12-01_05-19-41_2022-12-01_18-20-57        208   
1401  B208_2022-12-02_04-47-48_2022-12-02_19-40-01        208   
1405  B208_2022-12-07_05-13-02_2022-12-07_19-19-53        208   
1406  B208_2022-12-08_05-22-20_2022-12-08_18-39-15        208   

             startTime_iso  startTime_unix           endTime_iso  \
154   2020-03-03T04:42:38Z      1583210558  2020-03-03T19:44:51Z   
155   2020-03-06T04:53:23Z      1583470403  2020-03-06T19:44:42Z   
157   2020-03-09T14:16:13Z      1583763373  2020-03-09T19:34:17Z   
158   2020-03-10T04:50:03Z      1583815803  2020-03-10T19:51:25Z   
159   2020-03-12T04:56:41Z      1583989001  2020-03-12T19:44:57Z   
...                    ...             ...                   ...   
1399  2022-11-30T04:47:53Z      1669783673  2022-11-30T19:50:22Z   
1400  2022-12-01T05:19:41Z      1669871981  2022-12-01T18:20:57Z   
1401  2022-12-02T04:47:48Z      1669956468  2022-12-02T19:40:01Z   
1405  2022-12-07T05:13:02Z      1670389982  2022-12-07T19:19:53Z   
1406  2022-12-08T05:22:20Z      1670476940  2022-12-08T18:39:15Z   

      endTime_unix  drivenDistance busRoute  energyConsumption  \
154     1583264691       225047.90       83       1.544278e+09   
155     1583523882       224512.30       83       1.631816e+09   
157     1583782457        77824.36       83       5.406013e+08   
158     1583869885       225095.80       83       1.692171e+09   
159     1584042297       224181.20       83       1.145860e+09   
...            ...             ...      ...                ...   
1399    1669837822       223165.00       83       1.560888e+09   
1400    1669918857       190196.00       83       1.418847e+09   
1401    1670010001       224473.40       83       1.611150e+09   
1405    1670440793       210041.60       83       1.536697e+09   
1406    1670524755       190372.70       83       1.415700e+09   

      itcs_numberOfPassengers_mean  itcs_numberOfPassengers_min  \
154                       23.47531                            0   
155                       17.41578                            0   
157                       23.18182                            0   
158                       20.96410                            0   
159                       17.21235                            0   
...                            ...                          ...   
1399                      27.89066                            2   
1400                      26.03927                            0   
1401                      24.80384                            2   
1405                      28.78539                            0   
1406                      29.87940                            0   

      itcs_numberOfPassengers_max  status_gridIsAvailable_mean  \
154                           118                     0.472180   
155                            69                     0.451028   
157                            74                     0.460099   
158                            86                     0.475233   
159                            80                     0.340882   
...                           ...                          ...   
1399                          100                     0.456196   
1400                           96                     0.450413   
1401                           91                     0.438693   
1405                          115                     0.434858   
1406                          102                     0.439916   

      temperature_ambient_mean  temperature_ambient_min  

# 2. Extract all trips where `busRoute` is not a number.

### We get all the night busses.

In [7]:
df[df['busRoute'].str.contains(r'\D', na=False)]

name  busNumber  \
533   B183_2021-12-18_23-37-00_2021-12-19_03-38-35        183   
553   B183_2022-01-07_23-40-43_2022-01-08_03-31-21        183   
554   B183_2022-01-08_23-40-17_2022-01-09_03-35-32        183   
561   B183_2022-01-15_23-41-46_2022-01-16_03-40-23        183   
568   B183_2022-01-21_23-35-40_2022-01-22_03-26-24        183   
...                                            ...        ...   
1373  B208_2022-10-21_22-38-32_2022-10-22_02-42-21        208   
1374  B208_2022-10-22_22-34-45_2022-10-23_02-29-59        208   
1394  B208_2022-11-25_23-35-16_2022-11-26_03-30-39        208   
1407  B208_2022-12-09_23-55-12_2022-12-10_03-24-28        208   
1408  B208_2022-12-10_23-34-46_2022-12-11_03-29-05        208   

             startTime_iso  startTime_unix           endTime_iso  \
533   2021-12-18T23:37:00Z      1639870620  2021-12-19T03:38:35Z   
553   2022-01-07T23:40:43Z      1641598843  2022-01-08T03:31:21Z   
554   2022-01-08T23:40:17Z      1641685217  2022-01-09T03:35:32Z   
561   2022-01-15T23:41:46Z      1642290106  2022-01-16T03:40:23Z   
568   2022-01-21T23:35:40Z      1642808140  2022-01-22T03:26:24Z   
...                    ...             ...                   ...   
1373  2022-10-21T22:38:32Z      1666391912  2022-10-22T02:42:21Z   
1374  2022-10-22T22:34:45Z      1666478085  2022-10-23T02:29:59Z   
1394  2022-11-25T23:35:16Z      1669419316  2022-11-26T03:30:39Z   
1407  2022-12-09T23:55:12Z      1670630112  2022-12-10T03:24:28Z   
1408  2022-12-10T23:34:46Z      1670715286  2022-12-11T03:29:05Z   

      endTime_unix  drivenDistance busRoute  energyConsumption  \
533     1639885115        76216.06       N4        481350300.0   
553     1641612681        68557.06       N2        453625100.0   
554     1641699332        67962.92       N2        475383300.0   
561     1642304423        77156.70       N1        525168300.0   
568     1642821984        71917.75       N2        455476000.0   
...            ...             ...      ...                ...   
1373    1666406541        78567.16       N1        434776600.0   
1374    1666492199        73427.97       N2        399773700.0   
1394    1669433439        72911.26       N2        447553400.0   
1407    1670642668        59548.57       N1        451916500.0   
1408    1670729345        72293.83       N2        535095600.0   

      itcs_numberOfPassengers_mean  itcs_numberOfPassengers_min  \
533                       9.198582                            0   
553                       4.626984                            0   
554                       7.495495                            0   
561                       6.512500                            0   
568                       5.357143                            0   
...                            ...                          ...   
1373                     16.333330                            0   
1374                     17.710530                            0   
1394                     11.216670                            1   
1407                     20.105260                            0   
1408                     15.993200                            0   

      itcs_numberOfPassengers_max  status_gridIsAvailable_mean  \
533                            37                     0.491653   
553                            13                     0.427488   
554                            26                     0.515514   
561                            32                     0.473809   
568                            23                     0.493608   
...                           ...                          ...   
1373                           45                     0.431852   
1374                           57                     0.443358   
1394                           32                     0.465024   
1407                           74                     0.495739   
1408                           52                     0.459492   

      temperature_ambient_mean  temperature_ambient_min  

# 3. For each (busNumber, busRoute) pair, determine the number of trips.

In [8]:
df.groupby(['busRoute', 'busNumber'], as_index=False).size()

busRoute  busNumber  size
0        31        183    12
1        31        208     5
2        32        183    12
3        32        208    14
4        33        183   130
5        33        208    25
6        46        183   104
7        46        208    19
8        72        183   114
9        72        208    44
10       83        183   441
11       83        208   405
12       N1        183    10
13       N1        208     6
14       N2        183    19
15       N2        208    20
16       N4        183    11
17       N4        208     7

# 4. For each trip, compute the ratio between the energy consumption and the average number of passengers.

In [39]:
df['ratio'] = df['energyConsumption']/df['itcs_numberOfPassengers_mean']
df[['name', 'ratio']].head()

name         ratio
0  B183_2019-04-30_03-18-56_2019-04-30_08-44-20  8.640500e+07
1  B183_2019-04-30_13-22-07_2019-04-30_17-54-02  1.214747e+07
2  B183_2019-05-01_05-58-51_2019-05-01_22-32-30  7.342794e+07
3  B183_2019-05-03_02-50-21_2019-05-03_05-53-20  1.673328e+08
4  B183_2019-05-03_15-41-57_2019-05-03_23-06-24  2.613190e+07

# 5. For each station (`itcs_stopName`), determine the average number of passengers.

### First we have to concatenate all the 1408 CSV files in one data frame.
### To make the task less demanding for the CPU we select the only two columns we need for now.

In [58]:
df_stops = pd.DataFrame()
for name in df['name']:
    path = f'/Users/andreacardinali/Documents/ZTBus_compressed/{name}.csv'
    if os.path.exists(path):
        file = pd.read_csv(path, usecols=['itcs_numberOfPassengers', 'itcs_stopName'],
                                          na_values='-',
                                          dtype={'itcs_numberOfPassengers': float, 'itcs_stopName': str})
        df_stops = pd.concat([df_stops, file])

### Now we can determine the average number of passengers for each station.

In [52]:
df_stops.groupby('itcs_stopName', as_index=False)['itcs_numberOfPassengers'].mean().round(2)

itcs_stopName  itcs_numberOfPassengers
0                 Zürich, Albisrank                    16.20
1          Zürich, Albisriederplatz                    25.45
2         Zürich, Altes Krematorium                    21.98
3         Zürich, Bahnhof Affoltern                     2.85
4        Zürich, Bahnhof Altstetten                     7.82
..                              ...                      ...
144         Zürich, Zentrum Witikon                    17.26
145        Zürich, Zweiackerstrasse                     7.50
146              Zürich, Zwielplatz                     2.05
147             Zürich, Zwinglihaus                    23.20
148  Zürich,Kalkbreite/Bhf.Wiedikon                    23.16

[149 rows x 2 columns]

# 6. For each station, determine the buses that have stopped there at least once.

In [32]:
df_multistops = pd.DataFrame()
for name in df['name']:
    path = f'/Users/andreacardinali/Documents/ZTBus_compressed/{name}.csv'
    if os.path.exists(path):
        df_stops = pd.read_csv(path, na_values='-', usecols=['itcs_stopName'],
                               dtype={'itcs_stopName': str}).dropna(subset=['itcs_stopName'])
        df_stops['name'] = name
        df_multistops = pd.concat([df_multistops, df_stops])

In [33]:
df_stops3 = pd.merge(df_multistops, df, on='name')
df_completestops = df_stops3.groupby(['itcs_stopName','busNumber'], as_index=False).size()

In [42]:
df_completestops

itcs_stopName  busNumber   size
0                 Zürich, Albisrank        183   6324
1                 Zürich, Albisrank        208   6199
2          Zürich, Albisriederplatz        183  11218
3          Zürich, Albisriederplatz        208   7633
4         Zürich, Altes Krematorium        183  10022
..                              ...        ...    ...
292              Zürich, Zwielplatz        208      4
293             Zürich, Zwinglihaus        183   2791
294             Zürich, Zwinglihaus        208   1231
295  Zürich,Kalkbreite/Bhf.Wiedikon        183    279
296  Zürich,Kalkbreite/Bhf.Wiedikon        208    256

[297 rows x 3 columns]

# 7. For each station, determine the buses that have stopped there at least ten times.

In [35]:
df_completestops[df_completestops['size']>=10]

itcs_stopName  busNumber   size
0                 Zürich, Albisrank        183   6324
1                 Zürich, Albisrank        208   6199
2          Zürich, Albisriederplatz        183  11218
3          Zürich, Albisriederplatz        208   7633
4         Zürich, Altes Krematorium        183  10022
..                              ...        ...    ...
291              Zürich, Zwielplatz        183     77
293             Zürich, Zwinglihaus        183   2791
294             Zürich, Zwinglihaus        208   1231
295  Zürich,Kalkbreite/Bhf.Wiedikon        183    279
296  Zürich,Kalkbreite/Bhf.Wiedikon        208    256

[295 rows x 3 columns]

# 8. For each (route, bus) pair, compute the ratio between the overall energy consumption and the overall driven distance. 

In [72]:
df2 = df.groupby(['busNumber','busRoute'])[['drivenDistance','energyConsumption']].sum()
df2.reset_index(inplace=True)

In [73]:
df2['ratio'] = df2['energyConsumption']/df2['drivenDistance']
df2.round(2)

busNumber busRoute  drivenDistance  energyConsumption    ratio
0         183       31      2279952.00       1.387782e+10  6086.89
1         183       32      2256540.10       1.393320e+10  6174.59
2         183       33     29510481.41       1.761996e+11  5970.75
3         183       46     18224657.94       1.026237e+11  5631.04
4         183       72     23637238.45       1.394368e+11  5899.03
5         183       83     51249813.70       2.995540e+11  5844.98
6         183       N1       705122.71       4.217283e+09  5980.92
7         183       N2      1390618.35       7.929098e+09  5701.85
8         183       N4       766868.89       4.720068e+09  6154.99
9         208       31      1046306.68       5.535865e+09  5290.86
10        208       32      2337064.21       1.283340e+10  5491.25
11        208       33      6161892.39       3.474842e+10  5639.24
12        208       46      4186493.70       2.337366e+10  5583.11
13        208       72      9427138.15       5.100382e+10  5410.32
14        208       83     49965928.79       2.907837e+11  5819.64
15        208       N1       410691.93       2.316328e+09  5640.06
16        208       N2      1464008.67       7.913191e+09  5405.15
17        208       N4       519599.97       3.152512e+09  6067.19

# 9. Starting from the results of the previous point, for each route compute the buses with max and min energy ratio, and save the difference between these ratios in a dataframe.

In [77]:
groupmax = df2.groupby('busRoute')[['busNumber','ratio']].max()
groupmin = df2.groupby('busRoute')[['busNumber','ratio']].min()
df3 = pd.merge(groupmax, groupmin, on = 'busRoute').reset_index()
df3

busRoute  busNumber_x      ratio_x  busNumber_y      ratio_y
0       31          208  6086.891785          183  5290.862713
1       32          208  6174.585331          183  5491.247671
2       33          208  5970.747507          183  5639.244537
3       46          208  5631.035959          183  5583.110707
4       72          208  5899.029643          183  5410.318878
5       83          208  5844.977154          183  5819.638891
6       N1          208  5980.920966          183  5640.061883
7       N2          208  5701.850332          183  5405.152963
8       N4          208  6154.986937          183  6067.190112

In [78]:
df3['difference'] = df3['ratio_x'] - df3['ratio_y']
df3.head()

busRoute  busNumber_x      ratio_x  busNumber_y      ratio_y  difference
0       31          208  6086.891785          183  5290.862713  796.029072
1       32          208  6174.585331          183  5491.247671  683.337660
2       33          208  5970.747507          183  5639.244537  331.502969
3       46          208  5631.035959          183  5583.110707   47.925252
4       72          208  5899.029643          183  5410.318878  488.710765

# 10. Find the bus maximizing the difference computed in the previous point.

In [79]:
df3.loc[df3['difference'].idxmax()]

busRoute                31
busNumber_x            208
ratio_x        6086.891785
busNumber_y            183
ratio_y        5290.862713
difference      796.029072
Name: 0, dtype: object

# 11. Extract the rows of the details such that the `gnss_altitude` differs from the value in the preceding row. Store also the difference in the variable `altitude_variation`.

### To compute the difference of every line with the previous we can use the function .diff().
### Achtung: the first row of every file has no previous, so we just put a zero.

In [24]:
df_stops2['altitude_variation'] = df_stops2['gnss_altitude'].diff().fillna(0)

In [25]:
df_stops2[df_stops2['altitude_variation'] != 0]

time_iso   time_unix  electric_powerDemand  gnss_altitude  \
26     2019-06-24T03:16:39Z  1561346199                 0.000       393.1023   
27     2019-06-24T03:16:40Z  1561346200                -0.000       393.1976   
28     2019-06-24T03:16:41Z  1561346201                -0.000       393.2000   
32     2019-06-24T03:16:45Z  1561346205                -0.000       393.2976   
33     2019-06-24T03:16:46Z  1561346206                -0.000       393.3000   
...                     ...         ...                   ...            ...   
54096  2022-11-30T19:49:29Z  1669837769              7718.101       434.6953   
54097  2022-11-30T19:49:30Z  1669837770              7821.313       436.4188   
54098  2022-11-30T19:49:31Z  1669837771              7915.435       436.9681   
54099  2022-11-30T19:49:32Z  1669837772              2728.455       438.3276   
54100  2022-11-30T19:49:33Z  1669837773              2917.300       436.8831   

       gnss_course  gnss_latitude  gnss_longitude  itcs_busRoute  \
26        2.171190       0.826972        0.148489            NaN   
27        2.171190       0.826972        0.148489            NaN   
28        2.171190       0.826972        0.148489            NaN   
32        2.171190       0.826972        0.148489            NaN   
33        2.171190       0.826972        0.148489            NaN   
...            ...            ...             ...            ...   
54096     3.377910       0.826910        0.148451            NaN   
54097     3.383577       0.826904        0.148449            NaN   
54098     3.369137       0.826900        0.148448            NaN   
54099     3.349103       0.826896        0.148448            NaN   
54100     3.341115       0.826892        0.148447            NaN   

       itcs_numberOfPassengers itcs_stopName  ...  odometry_wheelSpeed_rr  \
26                         NaN           NaN  ...                     0.0   
27                         NaN           NaN  ...                     0.0   
28                         NaN           NaN  ...                     0.0   
32                         NaN           NaN  ...                     0.0   
33                         NaN           NaN  ...                     0.0   
...                        ...           ...  ...                     ...   
54096                      NaN           NaN  ...                     0.0   
54097                      NaN           NaN  ...                     0.0   
54098                      NaN           NaN  ...                     0.0   
54099                      NaN           NaN  ...                     0.0   
54100                      NaN           NaN  ...                     0.0   

       status_doorIsOpen  status_gridIsAvailable  status_haltBrakeIsActive  \
26                   0.0                     0.0                       0.0   
27                   0.0                     0.0                       0.0   
28                   0.0                     0.0                       0.0   
32                   0.0                     0.0                       0.0   
33                   0.0                     0.0                       0.0   
...                  ...                     ...                       ...   
54096                1.0                     0.0                       1.0   
54097                1.0                     0.0                       1.0   
54098                1.0                     0.0                       1.0   
54099                1.0                     0.0                       1.0   
54100                1.0                     0.0                       1.0   

       status_parkBrakeIsActive  temperature_ambient  traction_brakePressure  \
26                          1.0               296.15                246666.7   
27                          1.0               296.15                246666.7   
28                          1.0               296.15                246666.7   
32                          1.0               296.15       

# 12. For each details dataset, compute the sum of the absolute value (i.e. the sign is not considered) of `altitude_variation`.

In [26]:
df_stops4 = pd.DataFrame()
final = {}

for name in df['name']:
    path= f'/Users/andreacardinali/Documents/ZTBus_compressed/{name}.csv'
    if os.path.exists(path):
        df_stops = pd.read_csv(path, na_values='-', usecols=['gnss_altitude'], dtype={'gnss_altitude': float})
        df_stops['altitue_variation'] = df_stops['gnss_altitude'].diff().fillna(0)
        final[name] =  round(abs(df_stops['altitue_variation']).sum(),2)
        df_stops4 = pd.concat([df_stops4, df_stops])
        
final

{'B183_2019-06-24_03-16-13_2019-06-24_18-54-06': 24093.4,
 'B183_2020-07-19_03-04-41_2020-07-19_21-04-28': 27207.45,
 'B183_2021-07-26_03-53-27_2021-07-26_07-18-47': 4957.0,
 'B183_2021-10-05_13-53-20_2021-10-05_18-14-53': 6729.54,
 'B208_2021-05-27_04-09-18_2021-05-27_18-21-41': 8569.15,
 'B208_2021-09-25_05-59-40_2021-09-25_18-15-31': 7604.21,
 'B208_2022-07-16_05-30-37_2022-07-16_18-19-51': 9426.47,
 'B208_2022-10-09_03-26-03_2022-10-09_22-56-13': 14104.16,
 'B208_2022-10-26_03-25-19_2022-10-26_23-02-09': 14588.71,
 'B208_2022-11-30_04-47-53_2022-11-30_19-50-22': 9451.73}

# 13.  For each month of the year, compute the average ambient temperature.

In [10]:
for name in df['name']:
    path = f'/Users/andreacardinali/Documents/ZTBus_compressed/{name}.csv'
    if os.path.exists(path):
        lista = []

In [11]:
date = pd.to_datetime(df['startTime_iso'], format='mixed')
df['month'] = date.dt.month

In [13]:
temp = pd.DataFrame(df.groupby('month')['temperature_ambient_mean'].mean().round(2))
temp['Celsius'] = temp['temperature_ambient_mean'] - 273.15

In [14]:
temp.columns = ['Kelvin', 'Celsius']
temp

Kelvin  Celsius
month                 
1      278.96     5.81
2      280.53     7.38
3      283.32    10.17
4      286.61    13.46
5      290.05    16.90
6      296.14    22.99
7      297.38    24.23
8      295.85    22.70
9      292.36    19.21
10     287.71    14.56
11     281.88     8.73
12     279.38     6.23

# 14. For each bus compute the total time when the halt brake is active and the total time when the park brake is active. Compute also the ratio between those two times.

### I think we have to compute the sum of 'status_haltBrakeIsActive' and 'status_parkBrakeIsActive' for each CSV.
### Then we put both values in the smaller data frame.

In [ ]:
bella = {}
for name in df['name']:
    path = f'/Users/andreacardinali/Documents/ZTBus_compressed/{name}.csv'
    if os.path.exists(path):
        df_stops = pd.read_csv(path, usecols=['status_haltBrakeIsActive', 'status_parkBrakeIsActive'],
                               dtype={'status_haltBrakeIsActive': float, 'status_parkBrakeIsActive': float})
        haltbrake = df_stops['status_haltBrakeIsActive'].sum()
        parkbrake = df_stops['status_parkBrakeIsActive'].sum()
        bella[name] = (name, haltbrake, parkbrake)

In [197]:
bella_df = pd.DataFrame.from_dict(bella, orient='index', columns=['name', 'haltbrake', 'parkbrake']) # Convert the dictionary to a DataFrame
df = pd.merge(df,bella_df, on='name')

In [194]:
brake_time = df.groupby('busNumber')[['haltbrake', 'parkbrake']].sum()
brake_time['ratio'] = (brake_time['haltbrake']/brake_time['parkbrake']).round(2)
brake_time

haltbrake  parkbrake  ratio
busNumber                             
183        7837532.0  3181388.0   2.46
208        5942139.0  1828676.0   3.25

# 15. For each pair of stops that are consecutive in at least a trip, compute the average speed achieved when going from the first to the second stop.

In [15]:
for name in df['name']:
    path = f'/Users/andreacardinali/Documents/ZTBus_samples/{name}.csv'
    if os.path.exists(path):
        file = pd.read_csv(path, usecols=['itcs_stopName', 'odometry_vehicleSpeed'],
                                          na_values='-',
                                          dtype={'odometry_vehicleSpeed': float, 'itcs_stopName': str})
        i = 0
        print('\n')
        print(name)
        print('\n')
        row = file[file['itcs_stopName'].notna()]
        while i < len(row.index)-1:
            med = []
            for l in range(row.index[i], row.index[i+1]):
                med.append(file.iloc[l]['odometry_vehicleSpeed'])
            velo = sum(med)/len(med)
            print('From:', file.iloc[row.index[i]]['itcs_stopName'],'To:', file.iloc[row.index[i+1]]['itcs_stopName'], 'Speed:', round(velo,2), 'm/s or', round(velo*3.6,2), 'km/h')
            i += 1



B183_2019-06-24_03-16-13_2019-06-24_18-54-06


From: Zürich, Albisriederplatz To: Zürich, Hardplatz Speed: 3.36 m/s or 12.11 km/h
From: Zürich, Hardplatz To: Zürich, Bahnhof Hardbrücke Speed: 4.57 m/s or 16.46 km/h
From: Zürich, Bahnhof Hardbrücke To: Zürich, Schiffbau Speed: 8.58 m/s or 30.89 km/h
From: Zürich, Schiffbau To: Zürich, Escher-Wyss-Platz Speed: 7.95 m/s or 28.62 km/h
From: Zürich, Escher-Wyss-Platz To: Zürich, Rosengartenstrasse Speed: 5.67 m/s or 20.43 km/h
From: Zürich, Rosengartenstrasse To: Zürich, Bahnhof Wipkingen Speed: 5.05 m/s or 18.17 km/h
From: Zürich, Bahnhof Wipkingen To: Zürich, Nürenbergstrasse Speed: 7.17 m/s or 25.81 km/h
From: Zürich, Nürenbergstrasse To: Zürich, Rotbuchstrasse Speed: 7.14 m/s or 25.7 km/h
From: Zürich, Rotbuchstrasse To: Zürich, Schaffhauserplatz Speed: 3.71 m/s or 13.37 km/h
From: Zürich, Schaffhauserplatz To: Zürich, Scheuchzerstrasse Speed: 7.51 m/s or 27.03 km/h
From: Zürich, Scheuchzerstrasse To: Zürich, Seilbahn Rigiblick Speed: